worker/src/services/IngestionService/index.ts

```diff
@@ -252,61 +280,67 @@ export class IngestionService {
 
-    const recordsToMerge = clickhouseScoreRecord
-      ? [clickhouseScoreRecord, ...scoreRecords]
-      : scoreRecords;
+    // Set clickhouse first as this is the baseline for immutable fields
+    const recordsToMerge = [
+      clickhouseScoreRecord,
+      postgresScoreRecord,
+      ...scoreRecords,
+    ].filter(Boolean) as ScoreRecordInsertType[];
 
     const mergedRecord = this.mergeRecords(
       recordsToMerge,
       immutableEntityKeys[TableName.Scores],
     );
 
-    const parsedRecord = scoreRecordInsertSchema.parse(mergedRecord);
-    parsedRecord.event_ts = Math.max(
-      ...scoreRecords.map((r) => r.event_ts),
-      clickhouseScoreRecord?.event_ts ?? -Infinity,
-    );
-    return parsedRecord;
+    return scoreRecordInsertSchema.parse(mergedRecord);
   }
 
   private async mergeTraceRecords(params: {
     traceRecords: TraceRecordInsertType[];
+    postgresTraceRecord?: TraceRecordInsertType | null;
     clickhouseTraceRecord?: TraceRecordInsertType | null;
   }): Promise<TraceRecordInsertType> {
-    const { traceRecords, clickhouseTraceRecord } = params;
+    const { traceRecords, postgresTraceRecord, clickhouseTraceRecord } = params;
 
-    const recordsToMerge = clickhouseTraceRecord
-      ? [clickhouseTraceRecord, ...traceRecords]
-      : traceRecords;
+    // Set clickhouse first as this is the baseline for immutable fields
+    const recordsToMerge = [
+      clickhouseTraceRecord,
+      postgresTraceRecord,
+      ...traceRecords,
+    ].filter(Boolean) as TraceRecordInsertType[];
 
     const mergedRecord = this.mergeRecords(
       recordsToMerge,
       immutableEntityKeys[TableName.Traces],
     );
 
-    const parsedRecord = traceRecordInsertSchema.parse(mergedRecord);
-    parsedRecord.event_ts = Math.max(
-      ...traceRecords.map((r) => r.event_ts),
-      clickhouseTraceRecord?.event_ts ?? -Infinity,
-    );
-    return parsedRecord;
+    return traceRecordInsertSchema.parse(mergedRecord);
   }
 
   private async mergeObservationRecords(params: {
     projectId: string;
     observationRecords: ObservationRecordInsertType[];
+    postgresObservationRecord?: ObservationRecordInsertType | null;
     clickhouseObservationRecord?: ObservationRecordInsertType | null;
   }): Promise<ObservationRecordInsertType> {
-    const { projectId, observationRecords, clickhouseObservationRecord } =
-      params;
+    const {
+      projectId,
+      observationRecords,
+      postgresObservationRecord,
+      clickhouseObservationRecord,
+    } = params;
 
-    const recordsToMerge = clickhouseObservationRecord
-      ? [clickhouseObservationRecord, ...observationRecords]
-      : observationRecords;
+    // Set clickhouse first as this is the baseline for immutable fields
+    const recordsToMerge = [
+      clickhouseObservationRecord,
+      postgresObservationRecord,
+      ...observationRecords,
+    ].filter(Boolean) as ObservationRecordInsertType[];
```

Comment:
@maxdeichmann Regarding the domain timestamps: We always set clickhouse as the first record, postgres as second, and the any other observationRecord (sorted by event type and timestamp). As the domain timestamp is part of the immutableColumn list, we never overwrite what is written in the first entity. I.e. if we have a clickhouse record, we will deterministically use it, otherwise the domain timestamp from postgres and otherwise the one from the earliest event.

Based on my understanding this logic should cover our requrirements.

Reply:
Thanks!